lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,379 @@
+/// \file S3StorageClient.cpp
+/// \brief Implementation of StorageClient operations in S3
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <fstream>
+
+#include "llvm/Support/Process.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Storage/Path.h"
+#include "revng/Storage/S3StorageClient.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/OnQuit.h"
+#include "revng/Support/PathList.h"
+
+#include "aws/core/Aws.h"
+#include "aws/core/auth/AWSCredentials.h"
+#include "aws/core/utils/logging/FormattedLogSystem.h"
+#include "aws/s3/S3Client.h"
+#include "aws/s3/model/CopyObjectRequest.h"
+#include "aws/s3/model/DeleteObjectRequest.h"
+#include "aws/s3/model/GetObjectRequest.h"
+#include "aws/s3/model/HeadObjectRequest.h"
+#include "aws/s3/model/PutObjectRequest.h"
+
+using FilenameMapType = llvm::StringMap<std::string>;
+
+template<>
+struct llvm::yaml::CustomMappingTraits<FilenameMapType> {
+  static void inputOne(IO &IO, llvm::StringRef Key, FilenameMapType &Value) {
+    IO.mapRequired(Key.str().c_str(), Value[Key]);
+  }
+
+  static void output(IO &IO, FilenameMapType &Value) {
+    for (auto &Pair : Value) {
+      IO.mapRequired(Pair.first().str().c_str(), Pair.second);
+    }
+  }
+};
+
+namespace revng::detail {
+
+namespace {
+using Aws::Utils::Logging::FormattedLogSystem;
+using Aws::Utils::Logging::LogLevel;
+
+static Logger<> AWSLogger("aws");
+
+class LoggerSystem : public FormattedLogSystem {
+private:
+  std::mutex Mutex;
+
+public:
+  LoggerSystem(LogLevel LogLevel) : FormattedLogSystem(LogLevel) {}
+  ~LoggerSystem() override = default;
+  void Flush() override { AWSLogger.flush(); }
+
+  void ProcessFormattedStatement(Aws::String &&Statement) override {
+    // This function needs to be thread-safe, hence the lock_guard
+    std::lock_guard Guard(Mutex);
+    revng_log(AWSLogger, Statement);
+  }
+};
+
+class SDKInitImpl {
+private:
+  Aws::SDKOptions Options;
+
+public:
+  SDKInitImpl() : Options() {
+    // TODO: unix-specific, add #ifdef _WIN32 when needed
+    setenv("AWS_EC2_METADATA_DISABLED", "true", 1);
```

Comment:
What is this? Do we really have to `setenv`?

Reply:
> What is this?

When running the SDK inside an aws EC2 instance (a VM), there's a mechanism [described here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html) where the SDK, at init time, does some HTTP GETs to retrieve the credentials of the VM "user". This mechanism can only be disabled via the environment variable above

Reply:
This seems more like an hitman/orchestra concern to me. Shall we force it in orchestra?

What about this instead?
```c
// TODO: unix-specific, add #ifdef _WIN32 when needed
```

Reply:
> what if the container on K8s doesn't set the env?

It will be set by the `revng` wrapper script.

It's not a concern of `revng/revng` if we're running in a AWS EC2 or not, it's not the right place to set it.